### ISSUE TITLE
Cells get executed using different threads

### ISSUE THREAD

**Issue by dsyme** [dotnet/try#678]
#### Cells get executed using different threads

This is a design question and a possible bug.

As you execute different cells in a C# or F# notebook, a different thread gets used for each interaction.  The question is really whether we want to guarantee that the same thread gets used each time you execute a new cell.  Since cell execution is inherently linear it feels we should likely guarantee this.

For example, repeated re-execution of 
```csharp
System.Console.WriteLine("{0}", System.Threading.Thread.CurrentThread.ManagedThreadId);
```
gives 10, 11 etc.

This came up [when using TensorFlow.NET](https://github.com/SciSharp/TensorFlow.NET/issues/455), which had some thread local storage that was not getting re-initialized correctly as operations were performed from different threads.  But the question is more general too - is is something we want to guarantee or not, and we should document it either way.  

#### Please complete the following:
 - OS
    - [ ] Any
 - Browser 
    - [ ] Any

**Comment by jonsequitur** [comment]
Thinking more about this, I'm not seeing this as a bug so much as feedback pointing to two different needs we should address in the design:

1) People need a way to store data and share it across submissions. If you're writing an app and you understand your own threading model, thread-local storage or `AsyncLocal` or serialization or some other mechanism might variously be appropriate. But the notebook should abstract that in most cases, even to the degree that different submissions might execute in different processes or on different machines.

But also,

2) Reusing a thread is a legitimate need because of cases like the TensorFlow.NET example, so we should also allow people to control it. A magic command might be one way to approach this.

**Comment by dsyme** [comment]
Yup. I suspect you wouldn't lose anything if you commit to running cells on the same thread in the kernel.  Kernels are not designed to perform parallel/concurrent execution of notebook cells except where mediated by async/tasks etc, so there's no reason not to dedicate one thread in the kernel to serving these requests.  

The console F# Interactive does this for example. On windows each interaction is executed on a GUI thread suitable for Windows Forms or WPF programming, and on Linux we use a dedicated thread that can optionally go via a global "IEventLoop" held in the `fsi` object.

I'm actually a bit surprised that the `FSharp.Compiler.Service.dll` taht `dotnet-interactive` uses component doesn't also transfer across to the compilation/executor thread for the `Evaluate` calls being made into it. We should really do that at least for the compilation/checking part) to protect the compilation data structures from concurrent access.
